src/lib.rs

```diff
@@ -42,456 +42,417 @@
+    pub fn with_depth(mut self, depth: Depth) -> Self {
+        self.depth = depth;
+        self
+    }
+
+    pub fn with_auto_depth(mut self) -> Self {
+        self.set_auto_depth();
+        self
+    }
+
+    pub fn with_auto_reset(mut self, auto_reset: bool) -> Self {
+        self.set_auto_reset(auto_reset);
+        self
+    }
+
+    pub fn with_marker(mut self, marker: char) -> Self {
+        self.set_marker(marker);
+        self
+    }
+
+    pub fn set_depth(&mut self, depth: Depth) {
+        self.depth = depth;
+    }
+
+    pub fn set_auto_depth(&mut self) {
+        if let InferrenceResult::Color(depth) = Depth::try_inferr() {
+            self.depth = depth;
+        } else {
+            self.no_color = true;
+        }
+    }
+
+    pub fn set_auto_reset(&mut self, auto_reset: bool) {
+        self.auto_reset = auto_reset;
+    }
+
+    pub fn set_marker(&mut self, marker: char) {
+        self.marker = marker;
+        self.patterns = Patterns::new(marker);
+    }
+
+    /// Removes all Dahlia format codes from a string.
     ///
     /// ### Example
-    /// ```rs
-    /// let dahlia = Dahlia::new(Depth::High, true);
-    /// let text = dahlia.convert("&aHello\n&cWorld");
-    /// println!("{}", text);
+    /// ```rust
+    /// # use dahlia::{Dahlia};
+    /// let dahlia = Dahlia::default().with_auto_reset(false);
+    /// let green_text = "&2>be me";
+    /// assert_eq!(dahlia.clean(green_text), ">be me");
     /// ```
+    pub fn clean<'a>(&self, str: &'a str) -> Cow<'a, str> {
+        let converted = self.patterns.codes().replace_all(str, "");
+        self.finalize(converted)
+    }
+
+    /// Formats a string using the format codes.
     ///
-    /// <style>
-    /// .a {
-    ///     color: #55ff55;
-    /// }
-    /// .c {
-    ///     color: #ff5555;
-    /// }
-    /// </style>
-    /// <pre>
-    /// <span class="a">Hello</span>
-    /// <span class="c">World</span>
-    /// </pre>
-    pub fn convert(&self, string: String) -> String {
+    /// ### Example
+    /// ```rust
+    /// # use dahlia::{Dahlia, Depth};
+    /// let dahlia = Dahlia::default().with_depth(Depth::High);
+    /// let text = dahlia.convert("&aHello &cWorld");
+    /// assert_eq!(&text, "\x1b[38;2;85;255;85mHello \x1b[38;2;255;85;85mWorld\x1b[0m");
+    /// ```
+    pub fn convert<'a>(&self, str: &'a str) -> Cow<'a, str> {
         if self.no_color {
-            return clean(string, self.marker);
+            return self.clean(str);
         }
 
-        let reset = format!("{}r", self.marker);
+        let replacer = |captures: &Captures<'_>| self.get_ansi(captures);
+        let converted = self.patterns.codes().replace_all(str, replacer);
+        self.finalize(converted)
+    }
 
-        let string = if string.ends_with(&reset) || self.no_reset {
-            string
+    fn finalize<'a>(&self, str: Cow<'a, str>) -> Cow<'a, str> {
+        let str = if !str.ends_with("\x1b[0m") && self.auto_reset {
```

Comment:
Wouldn't it be better to check for auto_reset first?
```suggestion
        let str = if self.auto_reset && !str.ends_with("\x1b[0m") {
```

Reply:
This wasn't addressed yet it was resolved, any reason or just a mistake?

Reply:
The reason is that I am dumb and messed up when squashing the commit that fixed this

Reply:
Interestingly, it makes no difference in speed since both conditions compile to exactly one 64-bit number comparison (due to the string being exactly 4 ASCII chars). It does indeed look better the way you suggested though.

Reply:
Oh interesting, I sort of expected that but my Python-coded brain still believed it would make a difference 😄